src/Listeners/QueueJobSubscriber.php

```diff
@@ -0,0 +1,161 @@
+use Illuminate\Support\Arr;
+use Illuminate\Support\Collection;
+use JsonSerializable;
+use Psr\Container\ContainerInterface;
+use ReflectionClass;
+use ReflectionParameter;
+use Vinelab\Tracing\Contracts\ShouldBeTraced;
+use Vinelab\Tracing\Contracts\SpanContext;
+use Vinelab\Tracing\Facades\Trace;
+
+class QueueJobSubscriber
+{
+    /**
+     * @var ContainerInterface
+     */
+    protected $app;
+
+    /**
+     * QueueJobSubscriber constructor.
+     * @param  Container  $app
+     */
+    public function __construct(Container $app)
+    {
+        $this->app = $app;
+    }
+
+    /**
+     * @param  JobProcessing  $event
+     */
+    public function onJobProcessing(JobProcessing $event)
+    {
+        $this->startQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobProcessed  $event
+     */
+    public function onJobProcessed(JobProcessed $event)
+    {
+        $this->closeQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobFailed  $event
+     */
+    public function onJobFailed(JobFailed $event)
+    {
+        $this->closeQueueJobSpan($event);
+    }
+
+    /**
+     * @param  JobProcessing  $event
+     */
+    protected function startQueueJobSpan(JobProcessing $event)
+    {
+        /** @var Job $job */
+        $job = $event->job;
+        /** @var mixed $jobInstance */
+        $jobInstance = unserialize(Arr::get($job->payload(), 'data.command'));
+
+        if (in_array(ShouldBeTraced::class, class_implements($jobInstance))) {
+            $jobInput = $this->retrieveQueueJobInput($jobInstance);
+
+            $span = Trace::startSpan(class_basename($job->resolveName()), $jobInput->first(function ($attr) {
+                return $attr instanceof SpanContext;
+            }));
+
+            $span->tag('type', 'queue');
+            $span->tag('connection_name', $event->connectionName);
+            $span->tag('queue_name', $jobInstance->queue);
+            $span->tag('job_input', json_encode($this->normalizeQueueJobInputForLogging($jobInput)));
+
+            $this->app->instance('tracing.queue.span', $span);
+        }
+    }
+
+    /**
+     * @param JobProcessed|JobFailed $event
+     */
+    protected function closeQueueJobSpan($event)
+    {
+        if ($this->app->has('tracing.queue.span')) {
+            if ($event instanceof JobFailed) {
+                $this->app->get('tracing.queue.span')->tag('error', 'true');
+            }
+
+            $this->app->get('tracing.queue.span')->finish();
+
+            // If it's a sync driver, the main process will flush, we don't want to do that prematurely
+            if ($event->connectionName != 'sync') {
+                Trace::flush();
+            }
+        }
+    }
+
+    /**
+     * @param mixed $jobInstance
+     * @return Collection
+     */
+    protected function retrieveQueueJobInput($jobInstance): Collection
+    {
+        $class = new ReflectionClass($jobInstance);
+        $constructor = $class->getConstructor();
+
+        if ($constructor) {
```

Comment:
it would help for maintenance to know what the objective of this conditional here is.

Reply:
If the class doesn't have a constructor, `(new ReflectionClass($class))->getConstructor()` will return null. Actually, I think I will refactor this to `collect(optional($constructor)->getParameters())` and get rid of the condition entirely.